src/main/java/net/minecraftforge/common/data/JsonCodecProvider.java

```diff
@@ -0,0 +1,103 @@
+package net.minecraftforge.common.data;
+
+import java.io.IOException;
+import java.nio.file.Path;
+import java.util.Map;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+
+import com.google.gson.JsonElement;
+import com.mojang.serialization.Codec;
+import com.mojang.serialization.DynamicOps;
+import com.mojang.serialization.JsonOps;
+
+import cpw.mods.modlauncher.api.LamdbaExceptionUtils;
+import net.minecraft.core.Registry;
+import net.minecraft.core.RegistryAccess;
+import net.minecraft.data.CachedOutput;
+import net.minecraft.data.DataGenerator;
+import net.minecraft.data.DataProvider;
+import net.minecraft.resources.RegistryOps;
+import net.minecraft.resources.ResourceKey;
+import net.minecraft.resources.ResourceLocation;
+import net.minecraft.server.packs.PackType;
+import net.minecraftforge.common.data.ExistingFileHelper.ResourceType;
+import net.minecraftforge.forge.event.lifecycle.GatherDataEvent;
+
+/**
+ * <p>Dataprovider for using a Codec to generate jsons.
+ * Path names for jsons are derived from the given registry folder and each entry's namespaced id, in the format:</p>
+ * <pre>
+ * {@code <assets/data>/entryid/registryfolder/entrypath.json }
+ * </pre>
+ * 
+ * @param <T> the type of thing being generated.
+ * @param dataGenerator DataGenerator provided by {@link GatherDataEvent}.
+ * @param dynamicOps DynamicOps to encode values to jsons with using the provided Codec, e.g. {@link JsonOps.INSTANCE}.
+ * @param packType PackType specifying whether to generate entries in assets or data.
+ * @param directory String representing the directory to generate jsons in, e.g. "dimension" or "cheesemod/cheese".
+ * @param codec Codec to encode values to jsons with using the provided DynamicOps.
+ * @param entries Map of named entries to serialize to jsons. Paths for values are derived from the ResourceLocation's entryid:entrypath as specified above.
+ */
+public record JsonCodecProvider<T>(DataGenerator dataGenerator, ExistingFileHelper existingFileHelper, String modid, DynamicOps<JsonElement> dynamicOps, PackType packType, String directory, Codec<T> codec, Map<ResourceLocation, T> entries) implements DataProvider
```

Comment:
Making this a record means no one can extend it, which is potentially not what we want. Providing the frame while also allow extensibility sounds best to me.

Reply:
I would argue against your position here @SizableShrimp.

This API is stable and contains everything needed to handle the Codecs writing to disk.
If you want a custom handler here as a modder, I don't think this API is something for you anyway.

Leaving it a record is fine in my opinion.

Reply:
too late, it's already a definalized class and merged in forge through a separate PR